File: primes2.hs
--compile with ghc -Ox -threaded --make primes2.hs [where x is # of threads]
--run with ./primes2 +RTS -Nx [where x is the number from above]

--program that finds prime numbers
--multiprocessed

import Control.Monad
import Control.Parallel

innrprime :: Integer -> Integer -> [Bool]
innrprime n x =
        if ((x `mod` n) == 0) then
            [False]
        else
            [True]

prime :: Integer -> [[Bool]]
prime x = do
    let z = 2 :: Integer
    let w = (sqrt x + 1) :: Integer
    forM [z..w] $ \n -> innrprime n x

loop :: Integer -> Integer -> IO ()
loop x y =
    if (x > y) then
        return ()
    else
        if (False `elem` (head (prime x))) then
            loop (x+1) y
        else do
            print x
            loop (x+1) y

main :: IO ()
main = do
    putStrLn "Finding primes..."
    putStrLn "Enter max number to search to:"
    i <- getLine
    putStrLn "\n2"
    let y = read i :: Integer
    let x = 3 :: Integer
    loop x y

File: primes2.dart
/*
 * program that finds prime numbers
*/

import 'dart:math';
import 'dart:io';

void main() {
        print("");
        print("~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~");
        print(" This program finds prime numbers ");
        print("~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~");
        print("");

        print("Enter max number to search to:");

        // variable maxn is the maximum number that the program with search primes (starts at 2)
        int? maxn = int.tryParse(stdin.readLineSync()!);

        // check to see if input, attempt to make it into integer, else exit
        if(maxn == null) {
                print("Bad input");
                print('Press <Enter> to end program');
                String? _ = stdin.readLineSync()!;
                return;
        }

        // check if maxn is sane
        if(maxn <= 1) {
                print("Bad input");
                print('Press <Enter> to end program');
                String? _ = stdin.readLineSync()!;
                return;
        }

        print("The primes:");

        // find primes
        bool flag = false;
        for(int n = 2; n < maxn; n++) {
                flag = true;
                int sqrtn = int.parse((sqrt(n)+1).toString().split('.')[0]);
                for(int p = 2; p < sqrtn; p++) {
                        if(n % p == 0) {
                                flag = false;
                                break;
                        }
                }
                if(flag) {
                        print(n);
                }
        }

        print("Press <Enter> to end program");
        String? _ = stdin.readLineSync()!;
}
